grpclb_remote_balancer.go

```diff
@@ -0,0 +1,270 @@
+		lb.cc.UpdateBalancerState(lb.state, lb.picker)
+	}
+}
+
+// refreshSubConns creates/removes SubConns with backendAddrs. It returns a bool
+// indicating whether the backendAddrs are different from the cached
+// backendAddrs (whether SubConn was newed/removed).
+// Caller must hold lb.mu.
+func (lb *lbBalancer) refreshSubConns(backendAddrs []resolver.Address) bool {
+	lb.backendAddrs = nil
+	var backendsUpdated bool
+	// addrsSet is the set converted from backendAddrs, it's used to quick
+	// lookup for an address.
+	addrsSet := make(map[resolver.Address]struct{})
+	// Create new SubConns.
+	for _, addr := range backendAddrs {
+		addrWithoutMD := addr
+		addrWithoutMD.Metadata = nil
+		addrsSet[addrWithoutMD] = struct{}{}
+		lb.backendAddrs = append(lb.backendAddrs, addrWithoutMD)
+
+		if _, ok := lb.subConns[addrWithoutMD]; !ok {
+			backendsUpdated = true
+
+			// Use addrWithMD to create the SubConn.
+			sc, err := lb.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{})
+			if err == nil {
+				lb.subConns[addrWithoutMD] = sc // Use the addr without MD as key for the map.
+				lb.scStates[sc] = connectivity.Idle
+				sc.Connect()
+			} else {
+				grpclog.Warningf("roundrobinBalancer: failed to create new SubConn: %v", err)
+			}
+		}
+	}
+
+	for a, sc := range lb.subConns {
+		// a was removed by resolver.
+		if _, ok := addrsSet[a]; !ok {
+			backendsUpdated = true
+
+			lb.cc.RemoveSubConn(sc)
+			delete(lb.subConns, a)
+			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
+			// The entry will be deleted in HandleSubConnStateChange.
+		}
+	}
+
+	return backendsUpdated
+}
+
+func (lb *lbBalancer) readServerList(s *balanceLoadClientStream, done chan<- struct{}) {
+	defer close(done)
+	for {
+		reply, err := s.Recv()
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to recv server list: %v", err)
+			break
+		}
+		if serverList := reply.GetServerList(); serverList != nil {
+			lb.processServerList(serverList)
+		}
+	}
+}
+
+func (lb *lbBalancer) sendLoadReport(s *balanceLoadClientStream, interval time.Duration, done <-chan struct{}) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+		case <-done:
+			return
+		}
+		stats := lb.clientStats.toClientStats()
+		t := time.Now()
+		stats.Timestamp = &lbpb.Timestamp{
+			Seconds: t.Unix(),
+			Nanos:   int32(t.Nanosecond()),
+		}
+		if err := s.Send(&lbpb.LoadBalanceRequest{
+			LoadBalanceRequestType: &lbpb.LoadBalanceRequest_ClientStats{
+				ClientStats: stats,
+			},
+		}); err != nil {
+			grpclog.Errorf("grpclb: failed to send load report: %v", err)
+			return
+		}
+	}
+}
+
+func (lb *lbBalancer) watchRemoteBalancer() {
+	for {
+		select {
+		case <-lb.doneCh:
+			return
+		default:
+		}
+
+		lbClient := &loadBalancerClient{cc: lb.ccRemoteLB}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		stream, err := lbClient.BalanceLoad(ctx, FailFast(false))
```

Comment:
I think either `ctx` needs a deadline or we need to do this in a way that will unblock if lb.doneCh is closed:
```go
go func() { lbCh<-lbClient.BalanceLoad() }()
select {
  case s,err:=<-lbCh: ...
  case <-lb.doneCh:
    return
}
```

Reply:
When `lb.doneCh` gets closed (in `lbBalancer.Close()`), `lb.ccRemoteLB` is also closed.
So this RPC will fail in that case.

Reply:
In that case, can we suppress the error on line 187 if `doneCh` is closed?